src/analysis/bits-bounds-transfer-function.h

```diff
@@ -0,0 +1,367 @@
+        currElement.setUpperBound(std::max(0, maxBitsLeft - maxBitsRight + 1));
+        break;
+      }
+      case RemSInt64: {
+        if (right.constVal.has_value()) {
+          if (left.upperBound == 64) {
+            currElement.setUpperBound(64);
+          } else {
+            auto bitsRight =
+              Index(wasm::Bits::ceilLog2(right.constVal.value().geti64()));
+            currElement.setUpperBound(std::min(left.upperBound, bitsRight));
+          }
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case RemUInt64: {
+        if (right.constVal.has_value()) {
+          auto bitsRight =
+            Index(wasm::Bits::ceilLog2(right.constVal.value().geti64()));
+          currElement.setUpperBound(std::min(left.upperBound, bitsRight));
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case AndInt64: {
+        currElement.setUpperBound(std::min(left.upperBound, right.upperBound));
+        break;
+      }
+      case OrInt64:
+      case XorInt64: {
+        currElement.setUpperBound(std::max(left.upperBound, right.upperBound));
+        break;
+      }
+      case ShlInt64: {
+        if (right.constVal.has_value()) {
+          currElement.setUpperBound(
+            std::min(Index(64),
+                     Bits::getEffectiveShifts(right.constVal.value().geti64(),
+                                              Type::i64) +
+                       left.upperBound));
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case ShrUInt64: {
+        if (right.constVal.has_value()) {
+          auto shifts = std::min(Index(Bits::getEffectiveShifts(
+                                   right.constVal.value().geti64(), Type::i64)),
+                                 left.upperBound);
+          currElement.setUpperBound(
+            std::max(Index(0), left.upperBound - shifts));
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      case ShrSInt64: {
+        if (right.constVal.has_value()) {
+          if (left.upperBound == 64) {
+            currElement.setUpperBound(64);
+          } else {
+            auto shifts =
+              std::min(Index(Bits::getEffectiveShifts(
+                         right.constVal.value().geti64(), Type::i64)),
+                       left.upperBound);
+            currElement.setUpperBound(
+              std::max(Index(0), left.upperBound - shifts));
+          }
+        } else {
+          currElement.setUpperBound(64);
+        }
+        break;
+      }
+      default: {
+        addInformation = false;
+      }
+    }
+
+    if (collectingResults && addInformation) {
+      exprMaxBounds[curr] = currElement.upperBound;
+    }
+
+    currState->push(std::move(currElement));
+  }
+
+  void visitUnary(Unary* curr) {
+    MaxBitsLattice::Element unaryVal = currState->pop();
```

Comment:
```suggestion
    MaxBitsLattice::Element val = currState->pop();
```

No need to disambiguate between unary or anything else given that we're in `visitUnary`.